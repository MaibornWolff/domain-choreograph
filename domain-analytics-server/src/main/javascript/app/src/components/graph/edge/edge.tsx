import * as React from 'react';
import { pure } from 'recompose';
import { EDGE_ARROW_HEIGHT, EDGE_ARROW_WIDTH } from '~components/graph/graph';
import { SelectedOptions } from '~components/graph/models/selected-options';
import { Theme } from '~styles/theme';
import { Point } from '~types/point';
import { styled } from '~utils/styled';

function getEdgeColor({ theme, isSelected, selectedOptions = {} }: { theme: Theme } & EdgeProps): string {
  if (isSelected) {
    return selectedOptions.color || theme.colors.graph.selectedEdge;
  }
  return theme.colors.graph.edge;
}

const StyledPath = styled('path')<EdgeProps>`
  fill: none;
  stroke: ${getEdgeColor};
  stroke-width: 2px;
  transition: ${({ theme }) => theme.animations.all};
`;

const StyledPolygon = styled('polygon')<EdgeProps>`
  fill: ${getEdgeColor};
`;

export interface EdgeModel {
  id: string;
  line: string;
  outsideScope: boolean;
  points: Point[];
}

export interface EdgeProps {
  edge: EdgeModel;
  isSelected?: boolean;
  selectedOptions?: SelectedOptions;
}

export const _Edge: React.StatelessComponent<EdgeProps> = (props) => (
  <g>
    <StyledPath
      strokeDasharray={props.edge.outsideScope ? '3, 3' : undefined}
      d={props.edge.line}
      {...props}
    />
    {createRect(props)}
  </g>
);

export const Edge = pure(_Edge);

function createRect(props: EdgeProps) {
  const { edge } = props;
  const lastPoint: Point = edge.points[edge.points.length - 1];
  const secondLastPoint: Point = edge.points[edge.points.length - 2];
  const width = EDGE_ARROW_WIDTH;
  const height = EDGE_ARROW_HEIGHT;
  const x = lastPoint.x;
  const y = lastPoint.y;
  const angle = getAngleBeetweenTwoPointsInDegree(secondLastPoint, lastPoint) - 90;

  return <StyledPolygon
    x={x}
    y={y}
    width={width}
    height={height}
    points={`${x},${y} ${x - width / 2},${y - height} ${x + width / 2},${y - height}`}
    transform={`rotate(${angle} ${x} ${y})`}
    {...props}
  />;
}

function getAngleBeetweenTwoPointsInDegree(p1: Point, p2: Point): number {
  return Math.atan2(p2.y - p1.y, p2.x - p1.x) * 180 / Math.PI;
}
