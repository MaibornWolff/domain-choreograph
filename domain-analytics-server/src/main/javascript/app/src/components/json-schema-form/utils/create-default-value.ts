import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';

const defaultValues: Record<Exclude<JSONSchema4TypeName, 'object'>, any> = {
  string: undefined,
  number: undefined,
  integer: undefined,
  boolean: undefined,
  array: [],
  'null': null,
  any: ''
};

export function createDefaultValue(schema: JSONSchema4): any {
  const type = schema.type || 'any';
  if (type instanceof Array) {
    throw new Error('Array types are not supported');
  } else if (type === 'object') {
    return createDefaultValueForObject(schema);
  } else {
    return defaultValues[type];
  }
}

function createDefaultValueForObject(schema: JSONSchema4): Promise<object> {
  if (schema.type !== 'object') {
    throw new Error('Expect schema to be an object');
  }
  let defaultValue: any = {};
  if (schema.properties) {
    Object.entries(schema.properties).forEach(([key, property]) => {
      defaultValue[key] = createDefaultValue(property)
    });
  }
  return defaultValue;
}

