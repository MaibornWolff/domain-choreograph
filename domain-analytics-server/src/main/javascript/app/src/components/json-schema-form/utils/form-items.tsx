import { styled } from '~utils/styled';

export const FormItems = styled('div')`
    display: flex;
    flex-direction: column;

    & > * {
      margin-bottom: .5rem;
    }
`;
