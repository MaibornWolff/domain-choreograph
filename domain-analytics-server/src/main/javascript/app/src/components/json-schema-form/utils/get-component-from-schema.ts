import { FieldComponent } from 'clean-forms';
import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';
import { DatePickerMsField } from '~components/json-schema-form/form-fields/date-picker-ms-field';
import { NumberTextField, TextField } from '~components/json-schema-form/form-fields/text-field';
import { JsonSchemaFormFieldProps } from '~components/json-schema-form/models/json-schema-form-field-props';

export interface JsonSchemaComponentMap {
  matcher?: { type: JSONSchema4TypeName, format?: string };
  field: FieldComponent<any, JsonSchemaFormFieldProps>;
}

const defaultMapping: JsonSchemaComponentMap[] = [
  { matcher: { type: 'integer', format: 'utc-millisec' }, field: DatePickerMsField },
  { matcher: { type: 'integer' }, field: NumberTextField },
  { matcher: { type: 'number' }, field: NumberTextField },
  { field: TextField }
];

export function getComponentFromSchema(
  schema: JSONSchema4,
  mapping: JsonSchemaComponentMap[] = defaultMapping
): FieldComponent<any, JsonSchemaFormFieldProps> {
  for (const map of mapping) {
    if (map.matcher == null) {
      return map.field;
    }
    const typeIsMatching = map.matcher.type === schema.type;
    const formatIsMatching = map.matcher.format == null || map.matcher.format === schema.format;
    if (typeIsMatching && formatIsMatching) {
      return map.field;
    }
  }
  throw new Error('Couldn\'t match the schema');
}
