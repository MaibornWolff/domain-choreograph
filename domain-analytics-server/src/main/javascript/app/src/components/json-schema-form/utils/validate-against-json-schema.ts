import { ValidationFunction } from 'clean-forms';
import { JSONSchema4 } from 'json-schema';
import { Validator } from 'jsonschema';


export function validateAgainstJsonSchema(schema: JSONSchema4): ValidationFunction {
  const validator = new Validator();
  return ({ value }) => {
    const result = validator.validate(value, schema as any);
    const errors = result.errors.map(error => {
      let message = error.message;
      let property = error.property
        .replace(/^instance\.?/, '');
      let schema = error.schema as JSONSchema4;
      if (error.name === 'required') {
        message = `is required`;
        schema = schema.properties![error.argument];
        property = joinProperties(property, error.argument);
      }

      const title = schema.title || '';
      message = `${title} ${message}`;

      return [ property, message ] as [string, string];
    });
    return errors;
  }
}

function joinProperties(path1: string, path2: string): string {
  if (path1 === '') {
    return path2;
  }
  if (path2 === '') {
    return path1;
  }
  return `${path1}.${path2}`;
}
