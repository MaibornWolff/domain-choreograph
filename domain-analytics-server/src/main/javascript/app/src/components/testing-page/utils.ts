export function escapeJavaClass(javaClass: string) {
  return javaClass.replace(/\./g, '-');
}

export function reverseJavaClassEscape(javaClass: string) {
  return javaClass.replace(/-/g, '.');
}
