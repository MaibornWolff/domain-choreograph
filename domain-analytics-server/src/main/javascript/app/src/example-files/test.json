{
  "nodes": {
    "A": {
      "id": "A",
      "name": "A",
      "scope": "root"
    }
  },
  "dependencies": [],
  "scopes": {
    "root": {
      "id": "root",
      "executionContext": "Application",
      "nodes": [
        "A"
      ]
    }
  },
  "executionContexts": {
    "Application": {
      "id": "Application",
      "scopes": [
        "root"
      ]
    }
  }
}
