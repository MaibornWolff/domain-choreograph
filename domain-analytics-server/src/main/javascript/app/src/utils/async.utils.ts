export function wait(milliseconds: number): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    setTimeout(() => resolve(), milliseconds);
  });
}
