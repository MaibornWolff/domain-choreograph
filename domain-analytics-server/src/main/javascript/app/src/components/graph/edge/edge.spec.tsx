import * as React from 'react';
import { renderWithProviders } from '~utils/test.utils';
import { Edge, EdgeModel } from './edge';

describe('Edge', () => {
  it('should render inside scope', () => {
    const edge: EdgeModel = {
      id: 'a->b',
      line: 'M80,100L80,104.16666666666667C80,108.33333333333333,80,116.66666666666667,80,125C80,133.33333333333334,80,141.66666666666666,80,145.83333333333334L80,150',
      outsideScope: false,
      points: [{ x: 0, y: 0 }, { x: 1, y: 1 }]
    };
    expect(renderWithProviders(<Edge edge={edge}/>)).toMatchSnapshot();
  });

  it('should render outside scope', () => {
    const edge: EdgeModel = {
      id: 'a->b',
      line: 'M80,100L80,104.16666666666667C80,108.33333333333333,80,116.66666666666667,80,125C80,133.33333333333334,80,141.66666666666666,80,145.83333333333334L80,150',
      outsideScope: true,
      points: [{ x: 0, y: 0 }, { x: 1, y: 1 }]
    };
    expect(renderWithProviders(<Edge edge={edge}/>)).toMatchSnapshot();
  });

  it('should render if selected', () => {
    const edge: EdgeModel = {
      id: 'a->b',
      line: 'M80,100L80,104.16666666666667C80,108.33333333333333,80,116.66666666666667,80,125C80,133.33333333333334,80,141.66666666666666,80,145.83333333333334L80,150',
      outsideScope: false,
      points: [{ x: 0, y: 0 }, { x: 1, y: 1 }]
    };
    expect(renderWithProviders(<Edge edge={edge} isSelected={true}/>)).toMatchSnapshot();
  });
});
