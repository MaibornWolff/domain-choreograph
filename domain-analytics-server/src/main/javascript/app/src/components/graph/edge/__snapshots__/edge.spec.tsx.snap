// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edge should render if selected 1`] = `
.emotion-0 {
  fill: none;
  stroke: #657fcc;
  stroke-width: 2px;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
}

.emotion-2 {
  fill: #657fcc;
}

<g>
  <path
    class="emotion-0 emotion-1"
    d="M80,100L80,104.16666666666667C80,108.33333333333333,80,116.66666666666667,80,125C80,133.33333333333334,80,141.66666666666666,80,145.83333333333334L80,150"
  />
  <polygon
    class="emotion-2 emotion-3"
    height="8"
    points="1,1 -2.5,-7 4.5,-7"
    transform="rotate(-45 1 1)"
    width="7"
    x="1"
    y="1"
  />
</g>
`;

exports[`Edge should render inside scope 1`] = `
.emotion-0 {
  fill: none;
  stroke: #999;
  stroke-width: 2px;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
}

.emotion-2 {
  fill: #999;
}

<g>
  <path
    class="emotion-0 emotion-1"
    d="M80,100L80,104.16666666666667C80,108.33333333333333,80,116.66666666666667,80,125C80,133.33333333333334,80,141.66666666666666,80,145.83333333333334L80,150"
  />
  <polygon
    class="emotion-2 emotion-3"
    height="8"
    points="1,1 -2.5,-7 4.5,-7"
    transform="rotate(-45 1 1)"
    width="7"
    x="1"
    y="1"
  />
</g>
`;

exports[`Edge should render outside scope 1`] = `
.emotion-0 {
  fill: none;
  stroke: #999;
  stroke-width: 2px;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
}

.emotion-2 {
  fill: #999;
}

<g>
  <path
    class="emotion-0 emotion-1"
    d="M80,100L80,104.16666666666667C80,108.33333333333333,80,116.66666666666667,80,125C80,133.33333333333334,80,141.66666666666666,80,145.83333333333334L80,150"
    stroke-dasharray="3, 3"
  />
  <polygon
    class="emotion-2 emotion-3"
    height="8"
    points="1,1 -2.5,-7 4.5,-7"
    transform="rotate(-45 1 1)"
    width="7"
    x="1"
    y="1"
  />
</g>
`;
